Framework/Core/test/test_ASoAHelpers.cxx

```diff
@@ -842,10 +968,192 @@ BOOST_AUTO_TEST_CASE(BlockCombinations)
+    count++;
+  }
+  BOOST_CHECK_EQUAL(count, expectedUpperPairs.size());
+
+  // TODO: Change to sum of 3 table elements once it will be possible
+  expressions::Filter triplesFilter = (test::x > 4) && (test::x > test::x * (-1) + 9);
+  std::vector<std::tuple<int32_t, int32_t, int32_t>> expectedStrictlyUpperTriples{{7, 6, 4}, {7, 6, 5}};
+  std::vector<std::tuple<int32_t, int32_t, int32_t>> expectedUpperTriples{{7, 7, 3}, {6, 6, 4}, {7, 6, 4}, {7, 7, 4}, {5, 5, 5}, {6, 5, 5}, {7, 5, 5}, {6, 6, 5}, {7, 6, 5}, {7, 7, 5}, {6, 6, 6}, {7, 6, 6}, {7, 7, 6}, {7, 7, 7}};
+
+  count = 0;
+  for (auto& [t0, t1, t2] : selfCombinations<CombinationsFilteredFullIndexPolicy>(triplesFilter, testsA, testsA, testsA)) {
+    BOOST_CHECK_EQUAL(t0.x(), count % 3 + 5);
+    BOOST_CHECK_EQUAL(t1.x(), std::get<0>(expectedFullPairs[count / 3]));
+    BOOST_CHECK_EQUAL(t2.x(), std::get<1>(expectedFullPairs[count / 3]));
+    count++;
+  }
+  BOOST_CHECK_EQUAL(count, 3 * expectedFullPairs.size());
+
+  count = 0;
+  for (auto& [t0, t1, t2] : selfCombinations<CombinationsFilteredStrictlyUpperIndexPolicy>(triplesFilter, testsA, testsA, testsA)) {
+    BOOST_CHECK_EQUAL(t0.x(), std::get<0>(expectedStrictlyUpperTriples[count]));
+    BOOST_CHECK_EQUAL(t1.x(), std::get<1>(expectedStrictlyUpperTriples[count]));
+    BOOST_CHECK_EQUAL(t2.x(), std::get<2>(expectedStrictlyUpperTriples[count]));
+    count++;
+  }
+  BOOST_CHECK_EQUAL(count, expectedStrictlyUpperTriples.size());
+
+  count = 0;
+  for (auto& [t0, t1, t2] : selfCombinations<CombinationsFilteredUpperIndexPolicy>(triplesFilter, testsA, testsA, testsA)) {
+    BOOST_CHECK_EQUAL(t0.x(), std::get<0>(expectedUpperTriples[count]));
+    BOOST_CHECK_EQUAL(t1.x(), std::get<1>(expectedUpperTriples[count]));
+    BOOST_CHECK_EQUAL(t2.x(), std::get<2>(expectedUpperTriples[count]));
+    count++;
+  }
+  BOOST_CHECK_EQUAL(count, expectedUpperTriples.size());
+
+  std::vector<std::tuple<int32_t, int32_t, int32_t>> expectedStrictlyUpperTriples2{{7, 3, 0}, {6, 4, 0}, {7, 4, 0}, {6, 5, 0}, {7, 5, 0}, {7, 6, 0}, {7, 3, 1}, {6, 4, 1}, {7, 4, 1}, {6, 5, 1}, {7, 5, 1}, {7, 6, 1}, {7, 3, 2}, {6, 4, 2}, {7, 4, 2}, {6, 5, 2}, {7, 5, 2}, {7, 6, 2}, {6, 4, 3}, {7, 4, 3}, {6, 5, 3}, {7, 5, 3}, {7, 6, 3}, {6, 5, 4}, {7, 5, 4}, {7, 6, 4}, {7, 6, 5}};
+
+  count = 0;
+  for (auto& [t0, t1, t2] : selfCombinations<CombinationsFilteredStrictlyUpperIndexPolicy>(pairsFilter, testsA, testsA, testsA)) {
+    BOOST_CHECK_EQUAL(t0.x(), std::get<0>(expectedStrictlyUpperTriples2[count]));
+    BOOST_CHECK_EQUAL(t1.x(), std::get<1>(expectedStrictlyUpperTriples2[count]));
+    BOOST_CHECK_EQUAL(t2.x(), std::get<2>(expectedStrictlyUpperTriples2[count]));
+    count++;
+  }
+  BOOST_CHECK_EQUAL(count, 27);
+
+  // Order in tuple: (t0 > 5) && (t2 + t1 > 10) && (t4 + t3 > 9)
+  expressions::Filter fivesFilter = (test::x > 5) && (test::x > (test::x * (-1) + 9)) && (test::x > (test::x * (-1) + 4));
```

Comment:
Filter issues are most visible here:
- not possible to write e.g. `test::x + test::x > 9`, what makes filter hard to understand
- not intuitive order of iterators in a tuple due to tree structure of the filter. For example if `(test::x > 5)` would be the last condition, it still would be applied to `t0` (same order in tuple in spite of different order in filter expression)

Reply:
For this we said that the correct solution would be to have:

```C++
test::X{0} + test::X{1} > 9
```

where (0) and (1) refer to the first and the second table in your combination. This is still not there but I think it should be fairly straighforward, with test::x simply being an instance of `test::X{0}`.

Reply:
Yes, with {nr} that makes sense, if we wait with this PR until the filters are upgraded.